### ISSUE TITLE
Stabilize API

### ISSUE THREAD

**Issue by zepumph** [phetsims/faradays-law#107]
Throughout this issue, an unchecked box means" still needs to be completed for Faraday's Law 1.0". A checked box means "does not need to be completed for Faraday's Law 1.0" (whether deferred or completed).

@ariel-phet, @samreid and I met, and we took a lot of notes on the meeting, here is the list of work to be done. 

**General list to stabilize:**
- [x] studio (in progress)
- [x] SimiframeClient API: will be reviewed by @pixelzoom in https://github.com/phetsims/phet-io/issues/1366
- [x] Query parameter API: see https://github.com/phetsims/phet-io/issues/1389
- [x] static documentation: see https://github.com/phetsims/phet-io/issues/1390

**Notes/checklist**

- [x] https://github.com/phetsims/phet-io-website/issues/99 Make https://phet-io.colorado.edu/devguide/ more high level, focus on deploying documentation with each PhET-iO sim version.

- [x] **deferred** https://github.com/phetsims/phet-io-wrappers/issues/78 Should the wrapper JS be minified?  Right now we are thinking studio will be minified.  Sam and Michael will decide on the rest.  Michael says state wrapper and other smaller example wrappers should not be minified.  

- [x] https://github.com/phetsims/phet-io/issues/1325 phetio.invokeInputEvent is superseded by handlePlaybackEvent.js.  SR will create issue for eliminating invokeInputEvent.  SR thinks handlePlaybackEvent.js should be minified, but we can discuss it further.

- [x] https://github.com/phetsims/phet-io/issues/1324 In the studio, combine adjacent methods for PhET-iO.  OK to spend a small amount of time pairing related methods.

- [x] Put a timer on PhETIO.setInteractive/display.interactive = false to confirm the view still updates and the model still runs.  Update the documentation in scenery and PhETIO accordingly.

- [x] https://github.com/phetsims/phet-io-wrappers/issues/79 Investigate: `destinationClient.invoke( ‘phetio’, ‘launchSimulation’ );` vs `SimiframeClient.launchSim()`. Should clients only use `SimIFrameClient.launchSim()`? @samreid seemed to think there was a specific reason for using the invoke directly in state/mirror-inputs, but maybe it is vestigial. 

- [x] https://github.com/phetsims/phet-io-wrappers/issues/80 Consider adding a user interface for auto-invoke??  Could be added later without disturbing the API.

- [x] clean up startEvent/endEvent/trigger.  Make sure the wrapper types are clearly documented/examples, or provide a library for creating them.

- [x] https://github.com/phetsims/phet-io/issues/1324 lazyLink is before link in the type method api list, awkward.

- [x] https://github.com/phetsims/joist/issues/479 rename Screen.activeProperty to Screen.screenActiveProperty

- [x] Screen.activeProperty initial value false? Can we change that to true?

- [x] https://github.com/phetsims/joist/issues/481 Uninstrument Sim, and move the simStarted event and the ScreenShot method to phetio (PhETIO)

- [x] delete faradaysLaw.phetButtonPickableProperty as part of phetsims/scenery#711

- [x] the studio UI textbox should automatically update to changes of text in the simulation, then we can remove the "get" button, perhaps if the below comment comes through, then this applies to Property.<String> instead

- [x] https://github.com/phetsims/phet-io/issues/1322 There is no way for studio to know anything about FontIO, since it has no instances. It is instead used as a parameter to TextIO.setFontOptions( {{FontIO}} ). Same with Vector2.

- [x] https://github.com/phetsims/phet-io/issues/1326 Discussion with JO about the following:
* Text.js to use a property to store its text value. From a phet-io standpoint it looks like that would simplify code, because Text.js copies Property functions like `setText`/ `addListener`. Talked to JO about it.
* Jonathan's recommendation: decide on the api, and implement "property-like" types that use set/get/addListener to be consistent. Implementation shouldn't change, that is the job of the wrappers. Implement separate objects, and you can always display them differently in the studio (together a Text and it's textProperty).

In the meeting we made it through the studio to the Title Text Node. Although this is going slowly, @samreid and I both agree that good things are coming up in review/discussion.

I also don't think everything above needs to be completed before the publication of stable FL, as some 
is more cosmetic/documentation focused.

Tagging @kathy-phet and @ariel-phet so they are aware of the progress being done.

**Comment by zepumph** [comment]
> Investigate: destinationClient.invoke( ‘phetio’, ‘launchSimulation’ ); vs SimiframeClient.launchSim(). Should clients only use SimIFrameClient.launchSim()? @samreid seemed to think there was a specific reason for using the invoke directly in state/mirror-inputs, but maybe it is vestigial.

The reason why some wrappers use the more manual invocation is becuase the `onPhETiOInitialized` option passed into launch sim doesn't wait for a callback, so you can't make the async calls of (a) setting random seed and (b) setting playback mode into the launch sim option, because you run into a race condition between those and the launchSimulation async call. Either we keep it as is, or we could add those two phetio calls into the launchSim options, just like we do for others like `phetioEventsListener` and `simStartedMetadata`.

**Comment by zepumph** [comment]
>Screen.activeProperty initial value false? Can we change that to true?

I reviewed the implementation, and it looked safe to change the default, all screens are manually set in the sim's constructor.

**Comment by zepumph** [comment]
>the studio UI textbox should automatically update to changes of text in the simulation, then we can remove the "get" button, perhaps if the below comment comes through, then this applies to Property. instead

There is only a text changed listener in TextIO, but I think my implementation still makes sense in  https://github.com/phetsims/phet-io-wrappers/commit/1d2d55db470f281f1def0224c7eeb1e90be9faeb

**Comment by samreid** [comment]
> I reviewed the implementation, and it looked safe to change the default, all screens are manually set in the sim's constructor.

Does this mean that there is now a point during startup where multiple screens are active: true?  How does this change the data stream?

**Comment by zepumph** [comment]
In FAMB it adds these 4 events to the startup event sequence:
```
49 forcesAndMotionBasics.netForceScreen.activeProperty changed {"oldValue":true,"newValue":false}
50 forcesAndMotionBasics.motionScreen.activeProperty changed {"oldValue":true,"newValue":false}
51 forcesAndMotionBasics.frictionScreen.activeProperty changed {"oldValue":true,"newValue":false}
52 forcesAndMotionBasics.accelerationScreen.activeProperty changed {"oldValue":true,"newValue":false}
```

That's not ideal. I think it is better to have a single screen sim start "unactive" and change to active via startup, then add the above events. It makes more sense to me. You?

**Comment by samreid** [comment]
Notes from May 2, we can split these into separate issues later:

1. We noticed that it could be confusing to PhET-iO clients how the types are documented.  For instance, TextIO has a method `setFontOptions` which takes an argument of type `FontIO`.  A developer could reasonably think that they need to create `new FontIO(...)` in the wrapper and pass that to the `phetio.invoke` call.  However, that would be impossible since `FontIO` only exists in the simulation frame.  Therefore, we must make it clear that the client only passes primitives/functions/Objects (like literals), without client-side "types".  We must also document the rules for serializing/deserializing these, such as FontIO having `{weight:..., fontFamily:...`} etc.  For now, we will plan to include this documentation in the IO type `documentation`.

**Comment by zepumph** [comment]
- [ ] https://github.com/phetsims/chipper/issues/676 jsdoc isn't working right now.

**Comment by zepumph** [comment]
- [ ] **Deferred** naming collisions in studio when phetioID tails are named the same way (problem identified in https://github.com/phetsims/joist/issues/479 and potential solution in https://github.com/phetsims/phet-io-wrappers/issues/21#issuecomment-391078810)

@samreid I marked this as deferred, please undo this if you feel like it should be done before stable api.

**Comment by samreid** [comment]
That sounds OK for deferred since it does not affect stability of tandems.

**Comment by zepumph** [comment]
@samreid and I met again today. We discussed a few features of studio, and will continue on it more next week.

**Comment by zepumph** [comment]
Today we made a lot of progress. We went through from the beginning of the newly alphabetized studio and got to `faradaysLaw.faradaysLawScreen.view.flipMagnetButton` with https://github.com/phetsims/axon/issues/180.

We plan to continue next Wednesday the 25th.

**Comment by samreid** [comment]
I skimmed tandems below `flipMagnetButton` and nothing jumped out at me aside from https://github.com/phetsims/phet-io/issues/1345.  A closer look would still be helpful, I'll schedule it for Wednesday meeting with @chrisklus.

**Comment by samreid** [comment]
Molarity was recently published with master SHAs.  It did not have a "designed" phet-io API but it is using modern studio/API features, including many changes made for this issue.

I'm unassigning myself--this is still an important issue, but we will be working on it as a group in today's phet-io development meeting.

**Comment by zepumph** [comment]
today we got to the Reset All Button, but didn't look at it.

**Comment by chrisklus** [comment]
@zepumph and I have split the remaining tasks for this issue out into separate issues so there is nothing that remains in this one. I defer to @zepumph to close this issue.
